engine/src/state_chain_observer/client.rs

```diff
@@ -491,23 +524,77 @@ impl<BaseRpcClient: base_rpc_api::BaseRpcApi + Send + Sync + 'static, SignedExtr
 
 					// Often `finalized_header` returns a significantly newer latest block than the
 					// stream returns so we move the stream forward to this block
-					{
-						let finalised_header_hash =
-							base_rpc_client.latest_finalized_block_hash().await?;
-						let finalised_header =
-							base_rpc_client.block_header(finalised_header_hash).await?;
-						if finalized_block_stream.cache().number < finalised_header.number {
-							let blocks_to_skip =
-								finalized_block_stream.cache().number + 1..=finalised_header.number;
-							for block_number in blocks_to_skip {
+					let finalized_hash = base_rpc_client.latest_finalized_block_hash().await?;
+					let finalized_header = base_rpc_client.block_header(finalized_hash).await?;
+
+					let base_rpc_client = base_rpc_client.clone();
+					let mut finalized_block_stream: Pin<
+						Box<
+							dyn TryCachedStream<
+									Ok = BlockInfo,
+									Error = anyhow::Error,
+									Item = Result<BlockInfo, anyhow::Error>,
+								> + Send,
+						>,
+					> = match start_from {
+						Some(start_from) if start_from < finalized_block_stream.cache().number => {
+							// save this to avoid fetching the same block again
+							let latest_block_from_cache = *finalized_block_stream.cache();
+
+							// we can make this range exclusive because we've saved the cached block
+							// above, which we chain later.
+							let earlier_blocks_to_fetch =
+								start_from..finalized_block_stream.cache().number;
+
+							let base_rpc_client = base_rpc_client.clone();
+							let mut start_from_headers = Box::pin(
+								futures::stream::iter(earlier_blocks_to_fetch)
+									.then(move |block_number| {
+										let base_rpc_client_c = base_rpc_client.clone();
+										async move {
+											let hash = base_rpc_client_c
+												.block_hash(block_number)
+												.await?
+												.ok_or_else(|| {
+													anyhow::anyhow!(
+														"Block number {} does not exist in the state chain",
+														block_number
+													)
+												})?;
+											Ok(base_rpc_client_c.block_header(hash).await?)
+										}
+									})
+									.map_ok(|header| -> BlockInfo { header.into() }),
+							);
+
+							let start_block = start_from_headers
+								.next()
+								.await
+								.expect("start_from is less than block stream cache number")?;
+
+							Box::pin(
+								start_from_headers
+									.chain(futures::stream::once(futures::future::ready(Ok(
+										latest_block_from_cache,
+									))))
+									.chain(finalized_block_stream)
+									.make_try_cached(start_block),
+							)
+						},
+						_ => {
+							// No need to fetch any earlier blocks, we just have to move the stream
+							// to the correct starting point.
+							let skip_to = start_from.unwrap_or(finalized_header.number + 1);
```

Comment:
@AlastairHolmes this mimics the previous code, having the +1 effectively becomes the inclusive range. However, it's not clear to me that we actually want this. Skipping the inclusive range means that the next block that we pull on the stream will be the block *after* the header number - but if the header from the RPC call is the latest block, why would we want to do this, we're just delaying 6 seconds potentially?

Reply:
Pretty sure the old doesn't do that. This was something I thought about alot.

Reply:
Just to be clear, I think we should remove the +1, as that is how the old code works. Which having looking now, does appear to be the case.